llama/runner/cache.go

```diff
@@ -0,0 +1,148 @@
+package main
+
+import (
+	"log/slog"
+	"time"
+
+	"github.com/ollama/ollama/llama"
+)
+
+type TokenCache struct {
+	// context window size (per slot)
+	numCtx int
+
+	// individual caches
+	slots []TokenCacheSlot
+
+	lc *llama.Context
+}
+
+func NewTokenCache(lc *llama.Context, kvSize int, numSlots int) *TokenCache {
+	slots := make([]TokenCacheSlot, numSlots)
+
+	for i := range slots {
+		slots[i] = TokenCacheSlot{
+			id:     i,
+			tokens: make([]int, 0),
+		}
+	}
+
+	return &TokenCache{
+		numCtx: kvSize / numSlots,
+		slots:  slots,
+		lc:     lc,
+	}
+}
+
+type TokenCacheSlot struct {
+	// Index in the KV cache
+	id int
+
+	// tokens that are stored in the KV cache
+	tokens []int
+
+	// is this cache actively being processed as part of a sequence?
+	inUse bool
+
+	// last time this cache was used (as of start of processing)
+	lastUsed time.Time
+}
+
+func (t *TokenCache) LoadCacheSlot(prompt []int) (*TokenCacheSlot, []int, int) {
+	slot, numPast := t.findCacheSlot(prompt)
+
+	slot.inUse = true
+	slot.lastUsed = time.Now()
+
+	if numPast == len(prompt) {
+		// Leave one token to sample so we can get a response
+		numPast--
+	}
+
+	if !t.lc.KvCacheSeqRm(slot.id, numPast, -1) {
+		// Some models don't support partial erasure
+		t.lc.KvCacheSeqRm(slot.id, 0, -1)
+		numPast = 0
+	}
+
+	slog.Debug("loading cache slot", "id", slot.id, "cache", len(slot.tokens), "prompt", len(prompt),
+		"used", numPast, "remaining", len(prompt)-numPast)
+
+	prompt = prompt[numPast:]
+	slot.tokens = slot.tokens[:numPast]
+
+	return slot, prompt, numPast
+}
+
+func (t *TokenCache) findCacheSlot(prompt []int) (*TokenCacheSlot, int) {
```

Comment:
Similar to below, if there's a way to test this w/o the overhead of loading the llama context that would be awesome.

Reply:
I think it's a good idea - unfortunately, there is a call in that function that needs llama context and it's not that easy to work around. I ended up putting a nil check in findCacheSlot, which is not ideal but I do want to get more tests on the runner code and this a good spot.